### PR TITLE
[TD]Fix section snapping (fix #15961)

### LINKED ISSUE #15312: TechDraw: Snapping section views are failing when views are scaled.
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description
I appears that the snapping threshold is not taking into account the scaling of the views. So it is calculating a threshold that is much too big if scaling is something like 0.02.

Bug described here : https://forum.freecad.org/viewtopic.php?p=770197#p770197

### Full version info

```shell
0.22
```

### Subproject(s) affected?

Techdraw

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### LINKED ISSUE #15961: [TechDraw] Unable to move section along X axis
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

The section created on the basis of one of the views cannot be moved horizontally. No settings help.

At the same time, there are values ​​that allow you to move the section on the sheet. But the movement is reset if you move the section with the mouse.

Demonstration on video. 

https://github.com/user-attachments/assets/744acb10-4eab-4a64-92ce-e6f674d5cc7b

Remove the zip extension.
[Test_techDraw.FCStd.zip](https://github.com/user-attachments/files/16657276/Test_techDraw.FCStd.zip)

### Full version info

```shell
OS: Manjaro Linux (KDE/plasma)
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.38467 (Git) AppImage
Build type: Release
Branch: main
Hash: 87e73967da5734e100d5ac4b2fd5af1648a02572
Python 3.11.9, Qt 5.15.13, Coin 4.0.2, Vtk 9.2.6, OCC 7.7.2
Locale: Russian/Russia (ru_RU)
Installed mods: 
  * OpenTheme 2024.7.30
  * PitchedRoof


```

### Subproject(s) affected?

TechDraw

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by WandererFan**
This PR implements a fix for the section view snapping algo.  It also fixes #15312.

### CODE REVIEW

**Comment by PaddleStroke** [comment]
@WandererFan I have not tested it yet but I have two remarks: 
1 - In my PR I have also fixed the snapping of projection group items by adding in 2 places : 
```
        // if we are in a projection group, then we must take into account the group pos
        auto* parent = dynamic_cast<QGIView*>(view->parentItem());
        if (parent) {
            vPos = vPos + parent->pos();
        }
```
It is added in the section view snapping but also in the general view snapping. Please port it to your PR.

2 - Using an angle to snap is not ideal. Because then the threshold is not always the same. If you are very close to the parent view it is very hard to snap. If you are very far it becomes too easy.
Ideally for snapping you need a threshold that does not vary based on the distance between both views.

**Comment by WandererFan** [comment]
We could add an additional check on distance between the views, but the angle is the thing we are snapping to.    The distance factor could increase/decrease the angle limit, but I have no idea what reasonable values would be. 

I'll add the ProjectionGoup check for regular views.

**Comment by PaddleStroke** [comment]
> but the angle is the thing we are snapping to

Do you mean that the angle is the variable that is available to do the snapping? I have not studied the snapping mechanism you are using yet. But if you can snap using an angle, I would guess that you could snap using a distance as well.

image

**Comment by PaddleStroke** [comment]
From what I'm seeing it seems that you are using the center of the views to do the snapping. If so then it won't work in most cases because aligning the centers of the views is not correct in many cases. The section view will snap at the correct direction but not from the correct origin. ie the section view will be offset.

**Comment by PaddleStroke** [comment]
Ok so I can confirm that this new approach does not solve the oblique snapping : 
https://github.com/user-attachments/assets/c0510ebc-e3c1-454b-859a-8ea45e34fab4

It is actually having the same problem as I have had : the direction is correct but the origin is not.
But on top of that it has the downside of using angle instead of distance which have 2 negative effects : 
- Snapping threshold distance varies based on distance between views
- Snap only in one direction (though that could be fixed easily)
- Introduce unecessary new settings for the snap angle.

So long story short: if we are to solve the oblique snapping we need to be able to get the projection line of sectionOrigin on the section view. If you know how to do this then I can fix it using distance rather than angle.

**Comment by WandererFan** [comment]
<<Snap only in one direction (though that could be fixed easily)>>
I don't understand this one.  There is only 1 correct direction.  The section is to be aligned in the direction pointed to by the arrows on the section line.  Putting it on the other side of the base view would be wrong.

<<Introduce unecessary new settings for the snap angle.>>
We need a snapping on/off preference and the limit should be a preference rather than hard coded also.

**Comment by WandererFan** [comment]
If you are intent on using distances, then we don't need Preferences::SnapLimitSection().

To get a line from SectionOrigin on the base view to SectionOrigin on the section view,
use this method twice, then connect the points.

//! returns the position in the scene of a 3d point as it would appear in a QGIView.
//! The result is in Qt coordinates (inverted Y).
Base::Vector3d pointScenePos(Base::Vector3d point3d, QGIView* qgiv)
{
    auto viewFeature = qgiv->getViewObject();
    if (!viewFeature) {
        return Base::Vector3d();
    }
    auto pointInView = viewFeature->projectPoint(point3d);
    auto scenePos = qgiv->mapToScene(DU::toQPointF(pointInView));
    return DU::toVector3d(scenePos);
}

**Comment by WandererFan** [comment]
Missing a cast in that code.  

auto viewDvp = dynamic_cast<DrawViewPart*>(viewFeature);

**Comment by PaddleStroke** [comment]
Awesome, with this pointScenePos function it should work. I'm on it and will feedback how it goes.

**Comment by PaddleStroke** [comment]
image
Hmm yes it seems there's a problem, here it says origin on baseview is 0,0 which is not correct. I'm a bit confused with TD's various coordinates. Let's see if I can figure it out.

**Comment by PaddleStroke** [comment]
It seems to me that there is a missing mapping between the geometries and the view.

`    auto pointInView = dvp->projectPoint(Base::Vector3d(0.,0.,0.));`

is returning Base::Vector3d(0.,0.,0.) which seems sensible as it's also the origin of the projection. But then doing 

`    auto scenePos = qgiv->mapToScene(DrawUtil::toQPointF(pointInView));`

is mapping the point 0. 0. of the QGraphicsItem into the scene.

So yes it seems we are missing a mapping of the geometries on the view. 
Do you know how to do this?

**Comment by WandererFan** [comment]
The mapping is OK.  When a QGraphicsItem is created, it is assigned position (0,0), so if you try to map its position to the scene before a setPos() executes, you will get (0,0) as the scene position.  


Current code:
11:15:18  Loading Assembly workbench...
11:15:22  (qt5ct) D-Bus global menu: no
11:15:40  QGIV::itemChange - newPos:  (1169.05, -0.00) 
11:15:40  QGIV::pointScenePos - qgiv: View  scene: 1 parent: 0
11:15:40  QGIV::pointScenePos - qgiv.pos():  (0.00, 0.00) 
11:15:40  QGIV::pointScenePos - viewFeature.hasGeom: 0
11:15:40  QGIV::pointScenePos - qgiv: SectionView  scene: 1 parent: 0
11:15:40  QGIV::pointScenePos - qgiv.pos():  (0.00, 0.00)     <<<<
11:15:40  QGIV::pointScenePos - viewFeature.hasGeom: 0  <<<<
11:15:40  <Gui> Document.cpp(822): Cannot update representation for SectionOriginOnSectionView#SectionView  <<<<<

11:15:40  QGIV::itemChange - newPos:  (1169.05, -862.49) 
11:15:40  QGIV::pointScenePos - qgiv: View  scene: 1 parent: 0
11:15:40  QGIV::pointScenePos - qgiv.pos():  (0.00, 0.00) 
11:15:40  QGIV::pointScenePos - viewFeature.hasGeom: 0
11:15:40  QGIV::pointScenePos - qgiv: SectionView  scene: 1 parent: 0
11:15:40  QGIV::pointScenePos - qgiv.pos():  (0.00, 0.00) <<<<<<<
11:15:40  QGIV::pointScenePos - viewFeature.hasGeom: 0  <<<<<<<<<
11:15:40  QGIV::snapPosition - oBase: (0.000, 0.000)    <<<<<<<<<
11:15:40  QGIV::snapPosition - oSect: (0.000, 0.000)    <<<<<<<<<
11:15:40  mPos = (1169.047682, -862.492917)
11:15:40  oBase = (0.000000, 0.000000)
11:15:40  oSect = (0.000000, 0.000000)
11:15:40  dir = (-0.642788, 0.766044)
11:15:40  dist = 0.000000


11:15:40  QGIV::itemChange - newPos:  (1714.42, -1506.27) 
11:15:45  QGIV::itemChange - newPos:  (1169.05, -867.59) 
11:15:45  QGIV::pointScenePos - qgiv: View  scene: 1 parent: 0
11:15:45  QGIV::pointScenePos - qgiv.pos():  (1714.42, -1506.27) 
11:15:45  QGIV::pointScenePos - viewFeature.hasGeom: 1
11:15:45  QGIV::pointScenePos - qgiv: SectionView  scene: 1 parent: 0
11:15:45  QGIV::pointScenePos - qgiv.pos():  (1169.05, -862.49) <<<<<
11:15:45  QGIV::pointScenePos - viewFeature.hasGeom: 1          <<<<<
11:15:45  QGIV::snapPosition - oBase: (1714.417, -1506.274)     <<<<<
11:15:45  QGIV::snapPosition - oSect: (1169.048, -862.493)      <<<<<
11:15:45  mPos = (1169.047682, -867.590005)
11:15:45  oBase = (1714.416738, -1506.273521)
11:15:45  oSect = (1169.047682, -862.492917)
11:15:45  dir = (-0.642788, 0.766044)
11:15:45  dist = 3.962739

**Comment by WandererFan** [comment]
Code with check for hasGeometry:

11:24:26  Loading Assembly workbench...
11:24:30  (qt5ct) D-Bus global menu: no
11:24:44  QGIV::itemChange - newPos:  (1169.05, -0.00) 
11:24:44  QGIV::itemChange - newPos:  (1169.05, -862.49) 
11:24:44  QGIV::itemChange - newPos:  (1714.42, -1506.27) 
11:24:48  QGIV::itemChange - newPos:  (1174.14, -862.49) 
11:24:48  QGIV::pointScenePos - qgiv: View  scene: 1 parent: 0
11:24:48  QGIV::pointScenePos - qgiv.pos():  (1714.42, -1506.27) 
11:24:48  QGIV::pointScenePos - viewFeature.hasGeom: 1 
11:24:48  QGIV::pointScenePos - qgiv: SectionView  scene: 1 parent: 0
11:24:48  QGIV::pointScenePos - qgiv.pos():  (1169.05, -862.49) <<<<<
11:24:48  QGIV::pointScenePos - viewFeature.hasGeom: 1      <<<<<<
11:24:48  QGIV::snapPosition - oBase: (1714.417, -1506.274) <<<<<<
11:24:48  QGIV::snapPosition - oSect: (1169.048, -862.493)  <<<<<<
11:24:48  mPos = (1174.144770, -862.492917)
11:24:48  oBase = (1714.416738, -1506.273521)
11:24:48  oSect = (1169.047682, -862.492917)
11:24:48  dir = (-0.642788, 0.766044)
11:24:48  dist = 3.962739

**Comment by PaddleStroke** [comment]
Cannot get it to work on this test file : 
[section view.FCStd.zip](https://github.com/user-attachments/files/16336517/section.view.FCStd.zip)

**Comment by chennes** [comment]
@PaddleStroke can you give this another look now?

**Comment by PaddleStroke** [comment]
I can test on Monday.

**Comment by PaddleStroke** [comment]
Using the [file provided before](https://github.com/FreeCAD/FreeCAD/pull/15450#issuecomment-2243293153)
It appears there is still a problem with the alignment of the views : 
https://github.com/user-attachments/assets/23fa2dbf-6f34-4a1f-bd18-04b1abb2904f

**Comment by WandererFan** [comment]
Not seeing the problem in your video.  What am I missing?

**Comment by PaddleStroke** [comment]
@WandererFan here's a video with a clearer explanation:
https://github.com/user-attachments/assets/398d1697-b11b-4532-990c-916e5fc94e87

Let me know if it is clearer now

**Comment by WandererFan** [comment]
Finally figured this out.  Seems to work well on the sample file and others.

**Comment by WandererFan** [comment]
Needs a fix for projection group.

**Comment by maxwxyz** [comment]
Could you please add fixes #15961 I cannot link it
